jsonschema/reference.go

```diff
@@ -0,0 +1,168 @@
+						top.([]interface{})[intKey].(map[string]interface{})[key] = targetRef.(map[string]interface{})[key]
+						delete(top.([]interface{})[intKey].(map[string]interface{}), "$ref")
+					} else {
+						top.(map[string]interface{})[item].(map[string]interface{})[key] = targetRef.(map[string]interface{})[key]
+						delete(top.(map[string]interface{})[item].(map[string]interface{}), "$ref")
+					}
+				}
+			}
+		}
+	}
+
+	deref, err := json.Marshal(data)
+	return dj, deref, err
+}
+
+// walkInterface traverses the map[string]interface{} to located json references
+func walkInterface(node interface{}, source []string, refs []jsonRef) ([]jsonRef, error) {
+	var err error
+	for key, val := range node.(map[string]interface{}) {
+		switch reflect.TypeOf(val).Kind() {
+		case reflect.String:
+			if key == "$ref" {
+				refs = append(refs, jsonRef{
+					Source: source,
+					Target: val.(string),
+				})
+			}
+		case reflect.Slice:
+			for i, item := range val.([]interface{}) {
+				if reflect.TypeOf(item).Kind() == reflect.Map {
+					refs, err = walkInterface(item, append(source, key, strconv.Itoa(i)), refs)
+					if err != nil {
+						return nil, fmt.Errorf("unable to walk slice interface: %v", err)
+					}
+				}
+			}
+		case reflect.Map:
+			refs, err = walkInterface(node.(map[string]interface{})[key], append(source, key), refs)
+			if err != nil {
+				return nil, fmt.Errorf("unable to walk map interface: %v", err)
+			}
+		}
+	}
+	return refs, nil
+}
+
+// buildReference constructs the json reference: internal, file or http
+func buildReference(schemaPath string, top interface{}, ref string) (interface{}, error) {
+	target := strings.Split(ref, "#")
+	if len(target) < 2 {
+		target = append(target, "/")
+	}
+	var source interface{}
+
+	switch {
+	case len(target[0]) == 0:
+		source = top
+	case strings.HasPrefix(target[0], "http"):
+		res, err := goreq.Request{
+			Uri: target[0],
+		}.Do()
+		if err != nil {
+			return nil, fmt.Errorf("unable to get reference from %s: %v", target[0], err)
+		}
+		res.Body.FromJsonTo(&source)
+	default:
+		refPath, err := filepath.Abs(path.Dir(schemaPath) + "/" + target[0])
+		if err != nil {
+			return nil, fmt.Errorf("unable to expand reference filepath %s: %v", target[0], err)
+		}
+		data, err := ioutil.ReadFile(refPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read reference file %q: %v", refPath, err)
+		}
+		var dj schemaJSON
+		dj, data, err = Dereference(refPath, data)
+		if err != nil {
+			return dj, fmt.Errorf("failed to dereference refPath %s: %v", refPath, err)
+		}
+		json.Unmarshal([]byte(data), &source)
+	}
+	return parseReference(source, strings.Split(target[1], "/")[1:]), nil
+}
+
+// parseReference recursively parses the given reference path
+func parseReference(source interface{}, refPaths []string) interface{} {
+	if len(refPaths) > 1 {
+		return parseReference(source.(map[string]interface{})[refPaths[0]], refPaths[1:])
```

Comment:
Yeah, had thought of writing up a test for this --- just wasn't sure if I needed to test all the "non-public" (I forget the term) functions.  I originally had intended to write a test for this and so I will...

Reply:
There isn't a hard rule for unit tests when the functions like this are covered by other tests it is just the combination of recursion and empty interfaces that made me think it would benefit from that.

Reply:
Nice test, you should still check the type assertion though, it is quite unlikely to fail but if it does it will panic and so better to check it. Same for the other type assertion below. Something like
`sourceMap, ok := source.(map[string]interface{})
if !ok {//error}
`